test/discr/lp_discr_test.py

```diff
@@ -83,104 +75,66 @@ def test_init(exponent):
     with pytest.raises(ValueError):
         odl.DiscreteLp(space, part, rn_wrong_size)
 
 
-@skip_if_no_cuda
-def test_init_cuda(exponent):
-    # Normal discretization of unit interval
-    space = odl.FunctionSpace(odl.Interval(0, 1), out_dtype='float32')
-    part = odl.uniform_partition_fromintv(space.domain, 10)
-    rn = odl.CudaRn(10, exponent=exponent)
-    odl.DiscreteLp(space, part, rn, exponent=exponent)
-
-
-def test_factory(exponent):
-    discr = odl.uniform_discr(0, 1, 10, impl='numpy', exponent=exponent)
+def test_factory(exponent, fn_impl):
+    discr = odl.uniform_discr(0, 1, 10, impl=fn_impl, exponent=exponent)
 
-    assert isinstance(discr.dspace, odl.Fn)
+    assert isinstance(discr.dspace, FnBase)
+    assert discr.dspace.impl == fn_impl
     assert discr.is_rn
     assert discr.dspace.exponent == exponent
-    assert discr.dtype == odl.Fn.default_dtype(odl.RealNumbers())
 
     # Complex
-    discr = odl.uniform_discr(0, 1, 10, dtype='complex',
-                              impl='numpy', exponent=exponent)
-
-    assert isinstance(discr.dspace, odl.Fn)
-    assert discr.is_cn
-    assert discr.dspace.exponent == exponent
-    assert discr.dtype == odl.Fn.default_dtype(odl.ComplexNumbers())
-
+    try:
+        discr = odl.uniform_discr(0, 1, 10, dtype='complex',
+                                  impl=fn_impl, exponent=exponent)
 
-@skip_if_no_cuda
-def test_factory_cuda(exponent):
-    discr = odl.uniform_discr(0, 1, 10, impl='cuda', exponent=exponent)
-    assert isinstance(discr.dspace, odl.CudaFn)
-    assert discr.is_rn
-    assert discr.dspace.exponent == exponent
-    assert discr.dtype == odl.CudaFn.default_dtype(odl.RealNumbers())
-
-    # Cuda currently does not support complex numbers, check error
-    with pytest.raises(NotImplementedError):
-        odl.uniform_discr(0, 1, 10, impl='cuda', dtype='complex')
-
-
-def test_factory_dtypes():
-    real_float_dtypes = [np.float32, np.float64]
-    nonfloat_dtypes = [np.int8, np.int16, np.int32, np.int64,
-                       np.uint8, np.uint16, np.uint32, np.uint64]
-    complex_float_dtypes = [np.complex64, np.complex128]
-
-    for dtype in real_float_dtypes:
-        discr = odl.uniform_discr(0, 1, 10, impl='numpy', dtype=dtype)
-        assert isinstance(discr.dspace, odl.Fn)
-        assert discr.is_rn
-
-    for dtype in nonfloat_dtypes:
-        discr = odl.uniform_discr(0, 1, 10, impl='numpy', dtype=dtype)
-        assert isinstance(discr.dspace, odl.Fn)
-        assert discr.dspace.element().space.dtype == dtype
-
-    for dtype in complex_float_dtypes:
-        discr = odl.uniform_discr(0, 1, 10, impl='numpy', dtype=dtype)
-        assert isinstance(discr.dspace, odl.Fn)
+        assert isinstance(discr.dspace, FnBase)
+        assert discr.dspace.impl == fn_impl
         assert discr.is_cn
-        assert discr.dspace.element().space.dtype == dtype
+        assert discr.dspace.exponent == exponent
+    except TypeError:
```

Comment:
I don't think a wrong data type argument should raise a `TypeError`. It's technically wrong since the provided argument is not of the wrong type, but rather has the wrong value. If `ValueError` is too broad for this, maybe a custom error class would be preferable.